torch/optim/adadelta.py

```diff
@@ -187,9 +191,17 @@ def adadelta(
     See :class:`~torch.optim.Adadelta` for details.
     """
 
+    # We try to use the foreach implementation on CUDA whenever possible since
+    # it is faster than the for-loop implementation. However, the foreach
+    # implementation is not differentiable, so we must check differentiable=False.
+    # We still respect when the user inputs False for foreach.
     if foreach is None:
-        # Placeholder for more complex foreach logic to be added when value is not set
-        foreach = False
+        if not differentiable and all(
+            p.is_cuda for p in params + grads + square_avgs + acc_deltas
+        ):
+            foreach = True
+        else:
+            foreach = False
```

Comment:
nit
```suggestion
        foreach = not differentiable and all(
            p.is_cuda for p in params + grads + square_avgs + acc_deltas
        )
```
Also, another nit, you may prefer to simply chain the iterators of the lists rather than creating 3 temporary lists.

Reply:
Oh this is such a good point LOL thank you

Reply:
Never mind, we don't try itertools.chain because it does not play well with JIT. We get failures like:
```
RuntimeError:
2023-01-13T02:37:53.3718109Z 
2023-01-13T02:37:53.3718612Z chain(Tensor 0, Tensor 1, Tensor 2, Tensor 3) -> Tensor 0:
2023-01-13T02:37:53.3719214Z Expected a value of type 'Tensor' for argument '0' but instead found type 'List[Tensor]'.
2023-01-13T02:37:53.3720436Z Empty lists default to List[Tensor]. Add a variable annotation to the assignment to create an empty list of another type (torch.jit.annotate(List[T, []]) where T is the type of elements in the list for Python 2)
2023-01-13T02:37:53.3721315Z :
2023-01-13T02:37:53.3721949Z   File "/opt/conda/lib/python3.10/site-packages/torch/optim/adadelta.py", line 202
2023-01-13T02:37:53.3722429Z     if foreach is None:
2023-01-13T02:37:53.3723080Z         foreach = not torch.jit.is_scripting() and not differentiable and all(
2023-01-13T02:37:53.3723877Z             p.is_cuda for p in chain(params, grads, square_avgs, acc_deltas)
2023-01-13T02:37:53.3724932Z                                ~~~~~ <--- HERE
2023-01-13T02:37:53.3725478Z         )
```

I will use .extend though.